Fix #2645 - make cabal merge package preferences if more than one is specified

martinvlk opened the pull request:
Creating PR for this fix, need to discuss one thing before it is ready for merge. I will comment below.

martinvlk commented on cabal-install/Distribution/Client/Dependency.hs:
```diff
@@ -587,13 +587,19 @@ interpretPackagesPreference :: Set PackageName
+    versionPrefs = foldr mergeIfMultiplePrefs Map.empty prefs
+    mergeIfMultiplePrefs (PackageVersionPreference pkgname pref) m =
+      Map.insertWith (++) pkgname [pref] m
+    -- Previously the two cases below were simply ignored, and I think for a good
```
See the comment here - how do we handle "installed" or "latest" preference in the input pref list? Previously they were simply ignored by a partial list comprehension.

martinvlk commented on cabal-install/Distribution/Client/Dependency.hs:
```diff
@@ -587,13 +587,19 @@ interpretPackagesPreference :: Set PackageName
+    -- Previously the two cases below were simply ignored, and I think for a good
+    -- reason, because it is not possible to specify installed or latest preference
+    -- for a package. What to do?
+    mergeIfMultiplePrefs
```
See the comment here - how do we handle "installed" or "latest" preference in the input pref list? Previously they were simply ignored by a partial list comprehension.

grayjay replied:
`InstalledPreference` is handled below in `installPrefs`, so you can continue filtering it out here.  `interpretPackagesPreference` needs to handle both constructors because sandboxes add preferences on installed instances.

23Skidoo commented:
/cc @kosmikus

23Skidoo commented:
Travis failures are due to `-Werror`.

grayjay commented:
Do you mind if I review this pull request, too?

23Skidoo commented:
Sure, go ahead.

grayjay commented:
I'll add some advice and questions to the diff.  Most of the changes look fine.

grayjay commented on cabal-install/Distribution/Client/Dependency.hs:
```diff
@@ -818,7 +824,8 @@ resolveWithoutDependencies (DepResolverParams targets constraints
                            . packageId
-        versionPref   pkg = packageVersion pkg `withinRange` preferredVersions
+        versionPref   pkg = packageVersion pkg `withinRange`
+                            combinePrefRanges preferredVersions
```
Why does this function handle preferences differently than in the modular solver?  This seems more consistent:
```versionPref   pkg = length $ filter (packageVersion pkg `withinRange`) preferredVersions```

grayjay commented on cabal-install/Distribution/Client/Dependency/TopDown.hs:
```diff
@@ -127,7 +127,8 @@ explore pref (ChoiceNode _ choices)  =
-        isPreferred p = packageVersion p `withinRange` preferredVersions
+        isPreferred p = packageVersion p `withinRange`
+                        combinePrefRanges preferredVersions
         (PackagePreferences preferredVersions packageInstalledPreference)
```
The comment in `resolveWithoutDependencies` also applies here, and on the next change in this file.  I think that it would be better to either keep the old behavior or use the same comparison as in the modular solver, as long as the change isn't much larger.

grayjay commented on cabal-install/Distribution/Client/Dependency/Types.hs:
```diff
@@ -166,7 +168,11 @@ showPackageConstraint (PackageConstraintStanzas pn ss) =
 -- It is not specified if preferences on some packages are more important than
```
It would be good to update this comment and explain the meaning of multiple ranges.

grayjay commented on cabal-install/tests/UnitTests/Distribution/Client/Dependency/Modular/DSL.hs:
```diff
@@ -104,6 +107,7 @@ data ExampleDependency =
   | ExLang Language
 
+type ExPreference = (String, ExampleVersionRange)
 
```
I think that converting `ExPreference` into a data type would make the test cases easier to read.

grayjay commented on cabal-install/tests/UnitTests/Distribution/Client/Dependency/Modular/DSL.hs:
```diff
@@ -303,9 +307,9 @@ exResolve :: ExampleDb
+          -> (Bool, [ExPreference])
           -> ([String], Either String CI.InstallPlan.InstallPlan)
-exResolve db exts langs targets indepGoals = runProgress $
+exResolve db exts langs targets (indepGoals, prefs) = runProgress $
```
Why are independent goals and preferences combined in a pair?

grayjay commented on cabal-install/tests/UnitTests/Distribution/Client/Dependency/Modular/Solver.hs:
```diff
@@ -85,9 +88,25 @@ tests = [
+                         ,("A", mkvrOrEarlier [1,0,0])] $ mkTest db13 "selectPreferredVersionMultiple" ["A"] (Just [("A", 1)])
+        , runTest $ soft [("A", mkvrOrEarlier [1,0,0])
+                         ,("A", mkvrOrEarlier [2,0,0])] $ mkTest db13 "selectPreferredVersionMultiple2" ["A"] (Just [("A", 1)])
+        ]
```
I like the choice of test cases.

grayjay commented on cabal-install/tests/UnitTests/Distribution/Client/Dependency/Modular/Solver.hs:
```diff
@@ -85,9 +88,25 @@ tests = [
+mkvrThis = V.thisVersion . V.makeVersion 
+
+mkvrOrEarlier :: [Int] -> V.VersionRange
+mkvrOrEarlier = V.orEarlierVersion . V.makeVersion
```
You could simplify the versions in the test cases by making `mkvrThis` and `mkvrOrEarlier` both have type `ExamplePkgVersion -> V.VersionRange`.

martinvlk commented:
Hi, addressing review comments, all fair points - thanks!

martinvlk commented on cabal-install/Distribution/Client/Dependency/TopDown.hs:
```diff
@@ -669,8 +671,9 @@ finaliseSelectedPackages pref selected constraints =
         isPreferred p
-          | bounded   = True -- any constant will do
-          | otherwise = packageVersion p `withinRange` preferredVersions
+          | bounded   = 999 -- any constant will do
```
I am not sure here about this `bounded` case, what weight should we assign to it relative to the other preferences? Not sure I understand this bit.

grayjay replied:
I'm not familiar with the Topdown solver, but it looks like this function ignores preferred ranges when the dependency is bounded above.  Since `bounded` doesn't depend on the package instances being compared, the value here will only be compared with itself and should have no effect on behavior (which is consistent with the comment).  I would use something like `0` or `1`, so that it doesn't look like the constant has special meaning.  I guess moving the guard into `bestByPref` would remove the need for an arbitrary value, but I don't recommend refactoring it now that the solver is deprecated.

23Skidoo commented:
@grayjay Thanks for helping out with code review!

grayjay commented:
@23Skidoo Glad I can help.  I'm reviewing the latest set of changes now.

grayjay commented:
@martinvlk Thanks for making the changes.  I have a few more suggestions.

It looks like your first commit was combined with an existing commit, maybe because you rebased against the wrong branch.  Since you only modified code in the second commit, you can fix it by cherry-picking your new second commit on top of your original first commit, if you still have it.

grayjay commented on cabal-install/Distribution/Client/Dependency.hs:
```diff
@@ -587,13 +587,12 @@ interpretPackagesPreference :: Set PackageName
+      fromMaybe [anyVersion] (Map.lookup pkgname versionPrefs)
+    versionPrefs = foldr (\(PackageVersionPreference pkgname pref) m ->
+                            Map.insertWith (++) pkgname [pref] m)
+                   Map.empty prefs
```
This is a partial function now, and it could fail in a sandbox.  You could use a list comprehension and `Map.fromListWith`, since a pattern in a list comprehension acts as a filter.

grayjay commented on cabal-install/tests/UnitTests/Distribution/Client/Dependency/Modular/Solver.hs:
```diff
@@ -85,9 +88,22 @@ tests = [
+        , runTest $ soft [ ExPref "A" $ mkvrOrEarlier 1
+                         , ExPref "A" $ mkvrOrEarlier 2] $ mkTest db13 "selectPreferredVersionMultiple2" ["A"] (Just [("A", 1)])
+        ]
     ]
```
All of the test changes look good.  I just thought of another test case, if you want to add more.  Testing a preference on A==1 and a preference on A==2 would distinguish between the new behavior and merging preferences by taking the intersection of version ranges.

martinvlk commented:
Right so, I fixed the situation with wrongly squashed commits and also addressed the rest of the comments. BTW - thanks for them, this is helping me to bring my largely theoretical Haskell knowledge to real life - your patience is much appreciated!

This should now be ready for merging.

grayjay commented:
@martinvlk  No problem.  Thanks for fixing this issue!

The two new test cases have duplicate names.  Other than that, I think this pull request is ready to be merged.

martinvlk commented:
Ready for merge so!

grayjay commented:
Thanks!  It looks good.

BardurArantsson commented:
@grayjay Do you also want @kosmikus to review, or do think your sign-off is enough, i.e. do you think you have enough of a handle on the changes[1]? If you feel comfortable with it, I'm happy to press the "Merge" button.

[1] I'd assume you do, but you know what they say about assumptions...

grayjay commented:
@BardurArantsson I've reviewed the whole pull request, but I want to check with @23Skidoo before it is merged.

BardurArantsson commented:
A-okay :)

martinvlk commented:
Ping, any movement on this one?

BardurArantsson commented:
OK, I've reviewed this and even though I don't understand the solver in much detail, it all seems reasonable to me. (See comments.)

@23Skidoo Any objection to me merging this?

BardurArantsson commented:
Merged. Thanks for the contribution @martinvlk !
